src/scratch_impl.h

```diff
@@ -44,15 +54,18 @@ static size_t secp256k1_scratch_max_allocation(const secp256k1_scratch* scratch,
     return scratch->max_size - allocated - objects * ALIGNMENT;
 }
 
-static int secp256k1_scratch_allocate_frame(secp256k1_scratch* scratch, size_t n, size_t objects) {
+static int secp256k1_scratch_allocate_frame(const secp256k1_callback* error_callback, secp256k1_scratch* scratch, size_t n, size_t objects) {
     VERIFY_CHECK(scratch->frame < SECP256K1_SCRATCH_MAX_FRAMES);
 
-    if (n <= secp256k1_scratch_max_allocation(scratch, objects)) {
+    if (memcmp(scratch->magic, "scratch", 8) != 0) {
+        secp256k1_callback_call(error_callback, "invalid scratch space");
+        return 0;
+    }
+
+    if (n <= secp256k1_scratch_max_allocation(error_callback, scratch, objects)) {
         n += objects * ALIGNMENT;
```

Comment:
Isn't that too much? To store an array of `objects` things of type `t`, we need `objects * sizeof(t)` bytes of contiguous aligned memory. But there's no need to care about alignment here, because the stack frames are aligned already, no?

Reply:
We allocate multiple objects within one frame.

Reply:
Ah nevermind, I just understood the API...

I guess `objects * (ALIGNMENT - 1)` is still enough but that's not much of a difference.

Reply:
`(objects - 1) * ALIGNMENT` you mean :). I'm going to leave this be, because otherwise we get into thorny issues thinking about what happens when `objects == 0` and also we'd have to change it in a bunch of places. And it's only 16 bytes.

Reply:
I really meant `objects * (ALIGNMENT - 1)`. We're rounding up to a multiple of ALIGNMENT, so at most by `ALIGNMENT - 1`... But yes, just leave it, it's not worth the hassle.

Reply:
Oh! Yeah, I see. But did you further mean `(objects - 1) * (ALIGNMENT - 1)` since the first object does not require alignment (as the frame will already be aligned)?

I'll add a commit to do `ALIGNMENT-1` since that's quick and easy and doesn't create any new edge cases.

Reply:
> Oh! Yeah, I see. But did you further mean `(objects - 1) * (ALIGNMENT - 1)` since the first object does not require alignment (as the frame will already be aligned)?

Well, if the frame will already be aligned, then `(objects - 1) * (ALIGNMENT - 1)`  is fine and I think `objects == 0` should just be rejected. The problem is that now that we just have one malloc call, it's not guaranteed that the frame is aligned, not even the first one. (I think ROUND_TO_ALIGN from #566 helps here.)

Reply:
Oh, shit, yes, I think that's actually a bug in the current code. I need to add alignment when "allocating" frames as well.

Reply:
Fixed by eliminating frames.